lib/AnnotationToMetadataPass.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 The Clspv Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/IR/Constants.h"
+
+#include "AnnotationToMetadataPass.h"
+#include "Constants.h"
+
+using namespace llvm;
+
+PreservedAnalyses
+clspv::AnnotationToMetadataPass::run(Module &M, ModuleAnalysisManager &) {
+  for (auto &GV : M.globals()) {
+    if (GV.getName() == "llvm.global.annotations") {
+      // metadata node to write to
+      auto &context = M.getContext();
+      NamedMDNode *md_node =
+          M.getOrInsertNamedMetadata(clspv::EntryPointAttributesMetadataName());
+
+      // list of processed strings to delete at the end
+      SmallPtrSet<GlobalValue *, 4> to_erase;
+      // TODO: #816 remove after final transition.
+      SmallPtrSet<Constant *, 4> constants_to_erase;
+
+      ConstantArray *annotations_array =
+          dyn_cast<ConstantArray>(GV.getOperand(0));
+      for (auto &annotation_entry : annotations_array->operands()) {
+        ConstantStruct *annotation_struct =
+            dyn_cast<ConstantStruct>(annotation_entry.get());
+
+        auto op0 = annotation_struct->getOperand(0);
+        Function *entry_point = dyn_cast<Function>(
+            op0->getType()->isOpaquePointerTy() ? op0 : op0->getOperand(0));
```

Comment:
This PR pre-dates the switch to opaque pointers. Please update to avoid these tests.

Reply:
I have clean this file